db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/Predicate.java

```diff
@@ -521,34 +522,64 @@ public final boolean isStoreQualifier(){
      * @return true if the predicate is a pushable set of OR clauses.
      * @throws StandardException Standard exception policy.
      */
-    public final boolean isPushableOrClause(Optimizable optTable) throws StandardException{
+    public final boolean isPushableOrClause(Optimizable optTable, ConglomerateDescriptor cd,
+                                            boolean pushPreds) throws StandardException{
 
         if(andNode.getLeftOperand() instanceof OrNode){
-            QueryTreeNode node=andNode.getLeftOperand();
+            QueryTreeNode node = andNode.getLeftOperand();
+            QueryTreeNode trueNode;
+            AndNode tempAnd = null;
+
+            IndexRowGenerator irg = cd == null ? null : cd.getIndexDescriptor();
+            boolean isOnExpression = irg != null && irg.isOnExpression();
+            if (isOnExpression) {
+                trueNode = (QueryTreeNode) getNodeFactory().getNode(
+                        C_NodeTypes.BOOLEAN_CONSTANT_NODE,
+                        Boolean.TRUE,
+                        getContextManager());
+                tempAnd = (AndNode) getNodeFactory().getNode(
+                        C_NodeTypes.AND_NODE,
+                        trueNode,  // to be replaced later
+                        trueNode,
+                        getContextManager());
+            }
 
             while(node instanceof OrNode){
                 OrNode or_node=(OrNode)node;
 
                 if(or_node.getLeftOperand() instanceof RelationalOperator){
                     // if any term of the OR clause is not a qualifier, then
                     // reject the entire OR clause.
-                    if(!((RelationalOperator)or_node.getLeftOperand()).isQualifier(optTable,true)){
-                        // one of the terms is not a pushable Qualifier.
-                        return (false);
+                    if (isOnExpression) {
+                        tempAnd.setLeftOperand(or_node.getLeftOperand());
+                        tempAnd.postBindFixup();
+                        Predicate tempPred = (Predicate) getNodeFactory().getNode(
+                                C_NodeTypes.PREDICATE,
+                                tempAnd,
+                                or_node.getLeftOperand().getTablesReferenced(),
+                                getContextManager());
+                        boolean skipProbePreds = pushPreds && optTable.getTrulyTheBestAccessPath().getJoinStrategy().isHashJoin();
```

Comment:
pushPreds seems to be true when join planning is complete, and false in the middle of join planning.  If we use skipProbePreds as false while planning a hash join, then we might think an IN list predicate is useful, which may affect the cost of the join and lead us to pick it.  But then at the end of planning we would enter this code again, and see that we can't actually push the predicate, so the join is not as cheap as we thought.  Perhaps the code should be modified to detect during join planning whether the predicate can be pushed, something like:
> 
> JoinStrategy joinStrategy = pushPreds ? optTable.getTrulyTheBestAccessPath().getJoinStrategy() :
>                                                            optTable.getCurrentAccessPath().getJoinStrategy();
> boolean skipProbePreds = joinStrategy.isHashJoin();

Reply:
That makes total sense! I didn't really check this logic before but simply copy-pasted. I improve it as you suggested with a bit modification mainly because when join planning is complete, this method can also get called with `pushPreds = false`. This happens in some helper functions, e.g. `transferPredicates()`.